### PR TITLE
Add state, action, simulator and simple unit tests

### PR DESCRIPTION
**Pull request by ygreif**
## Background

Bellman's Folly is a tool that uses grid search to solve DSGE models.  A paper describing the a solution and a repo of grid search in multiple languages is provided here: https://github.com/jesusfv/Comparison-Programming-Languages-Economics http://economics.sas.upenn.edu/~jesusfv/comparison_languages.pdf

Essentially a DSGE is a maximization problem that looks like

sum_over_t B^t U(consumption)
s.t capital_next_period = F(capital this period) - consumption
more equations

So we want to solve for every level of capital and state of the world how much produce for next period to maximize utility.

One way to solve this from grid search.  We create a guess value function V which for a discretized parameter space maps to the lifetime value of being in that state. Then we

(1) Calculate the best new candidate P given V
(2) Use this new P to recalculate V
(3) Repeat until P, V converge

Currently, people either re-implement grid search.  Or they use Dynare which is a library if Matlab scripts.  Dynare is also ported to Octave.

### CODE REVIEW

File: model/action.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import abc
+
+
+class ActionSpace(object):
+    @abc.abstractmethod
+    def transition_probabilities(self, action, state):
```
Comment (ygreif) [inline model/action.py]: Add docstring

File: model/action.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import abc
```
Comment (ortutay) [inline model/action.py]: call the file action_space.py

File: model/state.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import abc
```
Comment (ortutay) [inline model/state.py]: call the file state_space.py

not sure what `abc` is? but I assume it's like a standard module or something?

File: tests/test_simulator.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    def transition_probabilities(self, action, state):
+        loc = state[1]
+        for delta, prob in [(-1, .5), (0, .25), (1, .25)]:
+            next_loc = min(max(loc + action + delta, 0), self.len - 1)
+            if 0 <= next_loc < self.len:
+                yield (0, next_loc), prob
+
+
+# TODO: make fixture or abandon py.test
```
Comment (ortutay) [inline tests/test_simulator.py:39]: yea I think you should use 
https://docs.python.org/2/library/unittest.html
and/or http://pythontesting.net/framework/nose/nose-introduction/

File: tests/test_simulator.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+s_s = SingleDimStateSpace(l)
+a_s = SingleDimActionSpace([0, 1, 2], valuations, l)
+value_func = numpy.asarray([[1.5, 2, 8, 12]])
+sim = simulator.Simulator(s_s, a_s, discount)
+
+
+def test_find_best_policy_for_state():
+    best_action, best_value = sim.find_best_policy_for_state((0, 0), value_func)
+    print best_action, best_value
+    assert best_action == 0
```
Comment (ortutay) [inline tests/test_simulator.py:53]: nit, be consistent about parens

File: model/state.py (hunk 3 above)
Comment (ygreif) [inline model/state.py]: It's for abstract base classes

File: tests/test_simulator.py (hunk 4 above)
Comment (ygreif) [inline tests/test_simulator.py:39]: nose isn't a unit testing framework.  It's a unit testing runner.  I find the syntax for the built in unit test awkward (self.assertEquals(arg1, arg2) vs assert arg1 == arg2. But I guess I'll live with it

File: tests/test_simulator.py (hunk 5 above)
Comment (ygreif) [inline tests/test_simulator.py:53]: ?

File: tests/test_simulator.py (hunk 4 above)
Comment (ygreif) [inline tests/test_simulator.py:39]: I think I'll just stick with py.test.  I think the debug messages for unittest are harder to read

File: simulator.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
+import numpy
+
+
+class Simulator(object):
+    def __init__(self, state_space, action_space, discount):
+        self.state_space = state_space
+        self.action_space = action_space
+        self.discount = discount
+        self.max_iterations = 1000
+        self.tolerance = 0.01
```
Comment (mikemintz) [inline simulator.py]: These seem like they should be constructor parameters, maybe with default values if you don't feel like setting them elsewhere.

File: simulator.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+    def find_best_policy_for_state(self, state, value_func):
+        best_action = None
+        best_payoff = float('-inf')
+        for a, payoff in self.action_space.get_actions(state):
+            for next_state, prob in self.action_space.transition_probabilities(a, state):
+                payoff += value_func[next_state] * prob * self.discount
```
Comment (mikemintz) [inline simulator.py]: It feels a bit icky to me to be mutating a variable from a for loop.

File: simulator.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        for state in self.state_space.states():
+            action, value = self.find_best_policy_for_state(state, value_func)
+            next_value_func[state] = value
+            next_policy_func[state] = action
+
+        return next_value_func, next_policy_func
+
+    def find_best_policy(self):
+        value_func = numpy.ndarray(self.state_space.dim())
+        for i in range(0, self.max_iterations):
```
Comment (mikemintz) [inline simulator.py]: Use `xrange` if in python 2. If python 3, `range` is good here

File: tests/test_simulator.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
+import numpy
+
+from model import state, action
+import simulator
+
+
+class SingleDimStateSpace(state.StateSpace):
+    def __init__(self, len):
```
Comment (mikemintz) [inline tests/test_simulator.py]: Although `len` isn't a keyword, it's an important built-in function that can be confusing if you override it with a local variable. I'd call this `length` or `size`

File: tests/test_simulator.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+    def dim(self):
+        return self.d
+
+    def states(self):
+        return [(0, i) for i in xrange(self.dim()[1])]
+
+
+class SingleDimActionSpace(action.ActionSpace):
+    def __init__(self, choices, valuations, len):
```
Comment (mikemintz) [inline tests/test_simulator.py]: Same note as above about `len` being a builtin function

File: tests/test_simulator.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+    def get_actions(self, state):
+        for choice in self.choices:
+            potential_action = choice + state[1]
+            if 0 <= potential_action < self.len:
+                yield (choice, -1 * choice * choice + self.valuations[state])
+
+    def transition_probabilities(self, action, state):
+        loc = state[1]
+        for delta, prob in [(-1, .5), (0, .25), (1, .25)]:
```
Comment (mikemintz) [inline tests/test_simulator.py]: Maybe extract this list to a variable with a name that makes it obvious why it has these seemingly arbitrary values

File: tests/test_simulator.py (hunk 5 above)
Comment (ortutay) [inline tests/test_simulator.py:53]: `assert best_action == 0`
(no parens)
vs

`assert (next_policy_func == [0, 1, 1, 0]).all()`
(has parens)

File: simulator.py (hunk 7 above)
Comment (ygreif) [inline simulator.py]: Isn't it common to accumulate values in a for loop?  Do you suggest using accumulate instead?

Comment (mikemintz) [inline simulator.py]: I'm not sure what you mean by accumulate? If it were me, I'd write

```
for a, payoff in self.action_space.get_actions(state):
    acc = payoff
    for next_state, prob in self.action_space.transition_probabilities(a, state):
        acc += value_func[next_state] * prob * self.discount
```

Maybe rename `acc` and `payoff` to correspond to what they're called in the algorithm, like `total_payoff` and `initial_payoff`

Comment (ygreif) [inline simulator.py]: Got it.  I thought it was the 'for' loop itself that you didn't like

File: simulator.py (hunk 6 above)
Comment (ygreif) [inline simulator.py]: Done.  Still not super pretty so we might revisit later

File: simulator.py (hunk 8 above)
Comment (ygreif) [inline simulator.py]: Should be done.  Removing the '0' from the range

File: tests/test_simulator.py (hunk 9 above)
Comment (ygreif) [inline tests/test_simulator.py]: Still need to address

**Pull request merged**